### ISSUE TITLE
Support Zstandard (zstd) compression algorithm

### ISSUE THREAD

**Issue by arzam16** [hashcat/hashcat#3220]
**Describe the feature**
Supporting the free/libre open source [Zstandard compression algorithm](https://github.com/facebook/zstd) will allow blazing fast on-the-fly access to the compressed dictionaries. While providing great compression rates especially for sorted data, [it is times faster](https://github.com/facebook/zstd#benchmarks) than xz (LZMA), gz and zip compression which are [already supported in hashcat](https://github.com/hashcat/hashcat/blob/3cacbe0e1f89301f3ae025969a0378af66f8a1f9/include/types.h#L1077).

**Current behavior**
No Zstandard support in the latest hashcat at the time of creating this FR. Currently I work it around by executing `zstdcat my_dictionary.txt.zst` and piping its output into hashcat running in the pipe input mode but it brings some drawbacks.

**Expected behavior**
Zstandard support for compressed password dictionaries.

**Hashcat version (please complete the following information):**
 - OS: Linux
 - Distribution: Debian
 - Version: 6.2.5

**Additional context**
The fastest password recovery suite and the most efficient modern compression algorithm are made for each other. Supporting "plain" zstd compression even without [its Dictionaries](https://github.com/facebook/zstd#dictionary-compression-how-to) option would be an awesome addition to hashcat.

**Comment by redongh** [comment]
I'd like to second the feature request to support on the fly decompression for [zstd](https://github.com/facebook/zstd) compressed wordlists as I also believe it is quite a good match for hashcat since it features good performance, efficiency and comes dual licensed (BSD, GPLv2).

### Some benchmarks

Ran some tests to better illustrate the performance gain practically achievable compared to gzip:
* compress wordlist kaonashi.dict (9.08 GiB)
   * zstd (default setting): `$ time zstd ./kaonashi.dict -o ./kaonashi.dict.zstd`
     * real: 1m12.937s, user 1m13.394s, sys: 0m2.482s
     * resulting size: 3.65 GiB
   * zstd (single-threaded): `$ time zstd --single-thread ./kaonashi.dict -o ./kaonashi.dict.zstd`
     * real: 1m26.579s, user 1m24.089s, sys: 0m2.230s
     * resulting size: 3.67 GiB
   * zstd (fast): `$ time zstd --fast ./kaonashi.dict -o ./kaonashi.dict.fast.zstd`
     * real: 0m34.813s, user 0m35.076s, sys: 0m2.585s
     * resulting size: 4.62 GiB
   * gzip (fast): `$ gzip -c --fast ./kaonashi.dict > ./kaonashi.dict.fast.gz`
     * real: 2m22.388s, user 2m20.716s, sys: 0m1.979s
     * resulting size: 3.68 GiB
   * gzip (default): `$ gzip -c ./kaonashi.dict > ./kaonashi.dict.gz`
     * real: 8m53.627s, user 8m50.000s, sys: 0m1.946s
     * resulting size: 3.22 GiB
 * decompress wordlist kaonashi.dict:
     * zstd (default) `time zstdcat ./kaonashi.dict.zstd > /dev/null`:
       * real: 0m13.158s, user 0m12.716s, sys: 0m0.404s
     * zstd (single-threaded) `$ time zstdcat --single-thread ./kaonashi.dict.zstd`:
       * real: 0m13.114s, user 0m12.732s, sys: 0m0.350s (no noticeable difference to non-single-treaded above)
     * zstd (single-threaded) `$ time zstdcat --single-thread ./kaonashi.dict.fast.zstd`:
       * real: 0m10.009s, user 0m9.543s, sys: 0m0.441s
     * gzip (fast) `time zcat ./kaonashi.dict.fast.gz > /dev/null`:
       * real: 1m5.806s, user 1m5.243s, sys: 0m0.399s
     * gzip (default) `time zcat ./kaonashi.dict.gz > /dev/null`:
       * real: 1m2.466s, user 1m1.990s, sys: 0m0.337s


#### Summary

* When comparing the **fast modes** of zstd vs gzip:
  * resulting filesize of the gzip-compressed  is about 20% smaller than the zstd-compressed one,
  * compression of the gzip-compressed wordlist takes ~ 4.09 times longer than the zstd-compressed one ,
  * decompression of the gzip-compressed wordlist takes ~ **6.58 times longer** than the zstd-compressed one. 

* When comparing the **default mode** of zstd vs the **fast mode** of gzip:
  * resulting filesizes are about equal (0.8% advantage on the side of zstd),
  * compression of the gzip-compressed wordlist takes ~ 1.95 times longer than the zstd-one,
  * decompression of the gzip-compressed wordlist takes ~ **5.00 times longer** than the zstd-one. 

* When comparing the **default modes** of zstd vs gzip:
  * resulting filesize of the gzip-compressed wordlist is about 11.9% smaller,
  * compression of the gzip-compressed wordlist takes ~ 7.32 times longer than the zstd-one,
  * decompression of the gzip-compressed wordlist takes ~ **4.75 times longer** than the zstd-one. 


### Additional notes

* All data measured on a relatively current notebook with an Intel-CPU; software versions used are gzip v1.11 and zstd v1.5.2.
* All times averaged on 3 subsequent runs.
* Resulting filesize parity to gzip is reached when applying compression level 8 on zstd; with this, zstd would take about half the time (4m0s) to compress the input file and decompress that output in 0m15.079s (less than a quarter of the gz-decompression time).

**Comment by tsweet64** [comment]
I also agree with this suggestion. I would love to have zstd support, since zstd is heavily asymmetric in favor of decompression speed. This would especially be useful when dealing with some >100GiB list files, which can be compressed highly with zstd and still decompress with very little load on the CPU.

**Comment by ghost** [comment]
Rather than net new code, reco dropping support for xz and gz in order to gain zstd. (net -1 feature to maintain).

Distros are migrating to zstd already.
https://archlinux.org/news/now-using-zstandard-instead-of-xz-for-package-compression/
https://www.phoronix.com/news/Linux-6.2-Zstd

**Comment by redongh** [comment]
somewhat similar to what I posted [earlier](https://github.com/hashcat/hashcat/issues/3220#issuecomment-1437313272) but in more illustrated way:

![240716_gz_vs_zstd_on_large_dictionary](https://github.com/user-attachments/assets/07ed885f-c67c-47f4-b6eb-eb5b63a36a59)

sidenote: compression levels were chosen in order to achieve a somewhat similar archive size.

**Comment by mohemiv** [comment]
Yeah, the idea is quite good and useful.

**Comment by elivoncoder** [comment]
i too think its a great idea and welcome zstd support!

**Comment by puppetma4ster** [comment]
Implementing the ability to read compressed zstd files would be really great. A must-have feature in 2026.

**Comment by jsteube** [comment]
zstd support has just been added https://github.com/hashcat/hashcat/commit/8371d3f
